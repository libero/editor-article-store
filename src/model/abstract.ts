import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import jsdom from "jsdom";

import { makeSchemaFromConfig } from './utils';
import * as abstractConfig from './config/abstract.config';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}
