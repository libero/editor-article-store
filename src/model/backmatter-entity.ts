import { v4 as uuidv4 } from 'uuid';
import {JSONObject} from "./types";

export abstract class BackmatterEntity {
  constructor() {
    this._id = uuidv4();
  }

  get id(): string {
    return this._id;
  }

  protected _id: string;

  protected abstract fromXML(xmlNode: Element): void;
  protected abstract fromJSON(json: JSONObject): void;
  protected abstract createBlank(): void;

  protected createEntity(data?: Element | JSONObject): void {
    if (!data) {
      this.createBlank();
<<<<<<< HEAD
    } else if (data instanceof Element) {
      this.fromXML(data);
    } else {
      this.fromJSON(data);
    }
  }
}
=======
    } else if (data.ownerDocument) {
      this.fromXML(data as Element);
    } else {
      this.fromJSON(data as JSONObject);
    }
  }
}
>>>>>>> origin/667-parse-related-articles
