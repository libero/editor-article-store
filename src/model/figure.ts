import { get } from 'lodash';

import { getTextContentFromPath } from './utils';

export interface Figure {
  copyrightHolder: string;
  copyrightStatement: string;
  copyrightYear: string;
  licenseType: string;
}

export const FIGURE_LICENSE_CC0 = 'LICENSE_CCO';
export const FIGURE_LICENSE_CC_BY_4 = 'LICENSE_CC_BY_4';
export const FIGURE_LICENSE_OTHER = 'LICENSE_OTHER';

const LICENSE_URLS_MAP: Record<string, string> = {
  'http://creativecommons.org/licenses/by/4.0/': FIGURE_LICENSE_CC_BY_4,
  'http://creativecommons.org/publicdomain/zero/1.0/': FIGURE_LICENSE_CC0
};

export function createFigureLicenseAttributes(el: Element): Figure {
  return {
    copyrightHolder: getTextContentFromPath(el, 'copyright-holder'),
    copyrightStatement: getTextContentFromPath(el, 'copyright-statement'),
    copyrightYear: getTextContentFromPath(el, 'copyright-year'),
    licenseType: getLicenseType(el)
  };
}

export function createEmptyLicenseAttributes(): Figure {
  return {
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: ''
  };
}

export function getLicenseUrl(licenseType: string): string {
  const licenseEntry = Object.entries(LICENSE_URLS_MAP).find(([type]) => type === licenseType);
  return licenseEntry ? licenseEntry[0] : '';
}

export function getFigureImageUrlFromXml(el: Element): string {
  return el.querySelector('graphic')!.getAttribute('xlink:href') || '';
}

function getLicenseType(el: Element): string {
  const licenseUrl =
    getTextContentFromPath(el, 'license_ref') || el.querySelector('license')!.getAttribute('xlink:href') || '';

  return LICENSE_URLS_MAP[licenseUrl] || FIGURE_LICENSE_OTHER;
}
