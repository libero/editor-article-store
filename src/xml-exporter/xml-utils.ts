import xmldom from "xmldom";
import nwmatcher from "nwmatcher";

export function parseXML(xml: string) {

  const xmlDoc = new xmldom.DOMParser().parseFromString(xml, 'text/xml');

  xmlDoc.constructor.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  xmlDoc.constructor.prototype.querySelectorAll = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).select(selectors, this);
  };

  const Element = xmlDoc.createElement('a').constructor;

  Element.prototype.matches = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).match(this, selectors);
  };

  Element.prototype.hasAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).hasAttribute(this, attributeName);
  };

  Element.prototype.getAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).getAttribute(this, attributeName);
  };

  Element.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  return xmlDoc;
}

export function clearNode(el: Element) {
  Array.from(el.childNodes).forEach(child => child.parentNode!.removeChild(child))
}
