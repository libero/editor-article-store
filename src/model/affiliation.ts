import { BackmatterEntity } from "./backmatter-entity";
import { JSONObject } from "./types";

export class Affiliation extends BackmatterEntity {
  label: string | undefined;
  institution: { name: string; } | undefined;
  address: { city: string; } | undefined;
  country: string | undefined;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }

  protected fromXML(_xml: Element): void {}

  protected fromJSON(json: JSONObject): void {
    this._id = (json._id as string) || this.id;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }
}