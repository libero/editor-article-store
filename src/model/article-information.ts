import {EditorState} from 'prosemirror-state';
import {DOMParser as ProseMirrorDOMParser} from 'prosemirror-model';
import {BackmatterEntity} from "./backmatter-entity";
import * as licenseTextConfig from './config/license-text.config';
import {getTextContentFromPath, makeSchemaFromConfig} from "./utils";
import {JSONObject} from "./types";
import {Person} from "./person";
import {get} from 'lodash';
import xmldom from "xmldom";

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

export class ArticleInformation extends BackmatterEntity {

  articleType: string = '';
  dtd: string = '';
  articleDOI: string = '';
  publisherId: string = '';
  volume: string = '';
  elocationId: string = '';
  subjects: Array<string> = [];
  publicationDate: string = '';
  licenseType: string = '';
  copyrightStatement: string = '';
  licenseText?: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (get(data, 'ownerDocument') && authors) {
      this.updateCopyrightStatementFromXml(data as Element, authors);
    }
  }

  public static createLicenseEditorState(content?: Element | null): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema
    });
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
            ? `${authors[0].lastName} et al`
            : '';

      const date = new Date(this.publicationDate);
      this.copyrightStatement = `Â© ${!isNaN(date.getTime()) ? date.getFullYear() : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    this.licenseText = EditorState.fromJSON({ schema }, json.licenseText as JSONObject);
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.ownerDocument.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') || '' : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="subject"] subject')).map(
      (el: Element) => el.textContent || ''
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }

    return '';
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        get(pubDateNode.querySelector('year'), 'textContent', ''),
        get(pubDateNode.querySelector('month'), 'textContent', ''),
        get(pubDateNode.querySelector('day'), 'textContent', ''),
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
