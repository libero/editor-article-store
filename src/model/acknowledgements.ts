import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { makeSchemaFromConfig } from './utils';
import xmldom from 'xmldom';
import {clearNode} from "../xml-exporter/xml-utils";
import {serializeManuscriptSection} from "../xml-exporter/manuscript-serializer";
import {Manuscript} from "../model/manuscript";

export function createAcknowledgementsState(content?: Element | null): EditorState {
  const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
  const ack = xmlContentDocument.createElement('ack');
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );

  if (content) {
    ack.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(ack),
    schema
  });
}

export function serializeAcknowledgementState(xmlDoc: Document, manuscript: Manuscript) {
  const acknowledgementsXml = serializeManuscriptSection(manuscript.acknowledgements, xmlDoc);
  let acknowledgementsEl = xmlDoc.querySelector('ack') as Element;

  if (!acknowledgementsEl) {
    acknowledgementsEl = xmlDoc.createElement('ack');
    acknowledgementsEl.setAttribute('id', 'ack');
    const ackTitleEl = xmlDoc.createElement('title');

    ackTitleEl.appendChild(xmlDoc.createTextNode('Acknowledgements'));
    acknowledgementsEl.appendChild(ackTitleEl);
    xmlDoc.querySelector('back')?.appendChild(acknowledgementsEl);
  }

  const ackTitleEl = acknowledgementsEl.firstChild as Element;
  clearNode(acknowledgementsEl);
  acknowledgementsEl.appendChild(ackTitleEl);
  acknowledgementsEl.appendChild(acknowledgementsXml);
}
