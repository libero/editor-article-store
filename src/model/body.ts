import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';

import * as bodyConfig from './config/body.config';
import { makeSchemaFromConfig } from './utils';
import xmldom from 'xmldom';
import {Manuscript} from "./manuscript";
import {serializeManuscriptSection} from "../xml-exporter/manuscript-serializer";
import {clearNode} from "../xml-exporter/xml-utils";

export function createBodyState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
  const body = xmlContentDocument.createElement('body');
  if (content) {
    body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(body),
    schema
  });
}

export function serializeBodyState(xmlDoc: Document, manuscript: Manuscript) {
  const bodyXml = serializeManuscriptSection(manuscript.body, xmlDoc);
  const bodyEl = xmlDoc.querySelector('body') as Element;

  clearNode(bodyEl);
  bodyEl.appendChild(bodyXml);
}
