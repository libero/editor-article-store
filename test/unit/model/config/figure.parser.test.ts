import * as jsdom from "jsdom";

import {parseFigure, serializeFigure} from '../../../../src/model/config/figure.parser';
import {createBodyState} from '../../../../src/model/body';
import {parseXML} from "../../../../src/xml-exporter/xml-utils";

const FIGURE_CONTENT_XML = `<fig id="fig2">
  <label>Figure 1.</label>
  <caption>
    <title>Bacterial cells on the surface of a plant leaf.</title>
    <p>
        Pellentesque porta interdum sodales (<xref ref-type="bibr" rid="bib6">Heissler and Sellers, 2016a</xref>). 
        Vestibulum<sub>id</sub> interdum<sup>felis</sup>. 
        <bold>Vestibulum</bold> at <underline>viverra</underline> lorem. <sc>Aliquam Gravida</sc> nulla orci, quis tristique tortor ornare at 
        <ext-link ext-link-type="uri" xlink:href="https://www.biobam.com/">https://www.biobam.com/</ext-link>. 
        Pellentesque accumsan sed ipsum eget hendrerit (<xref ref-type="bibr" rid="bib3">Foley et al., 2018</xref>). 
        Pellentesque cursus, nibh eu <monospace>mattis condimentum</monospace>, massa libero faucibus enim, 
        et tempor orci neque eget<sup>mi</sup>. Phasellus lacus<sub>mi</sub>, ultrices quis placerat id, dignissim ac erat.
    </p>
  </caption>
  <graphic mime-subtype="jpg" mimetype="image" xlink:href="fig1.jpg"/>
</fig>`;

jest.mock('../../../../src/model/figure', () => ({
  getFigureImageUrlFromXml: () => 'IMAGE_URL',
  createFigureLicenseAttributes: () => ({
    copyrightHolder: 'Copyright holder',
    copyrightStatement: 'Copyright stmt',
    copyrightYear: 'Copyright year',
    licenseType: 'License type'

  }),
  createEmptyLicenseAttributes: () => ({
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: ''
  })
}));

describe('figure parsing spec', () => {
  it('checks empty figure parsing', () => {
    const dom = new jsdom.JSDOM();
    const body = dom.window.document.body;
    const figure = dom.window.document.createElement('fig');
    body.appendChild(figure);
    const bodyState = createBodyState(dom.window.document.body);
    expect(parseFigure(body.firstChild as Element, bodyState.schema)).toMatchSnapshot();
  });

  it('checks figure parsing', () => {
    const dom = new jsdom.JSDOM();
    const bodyState = createBodyState(dom.window.document.createElement('body'));
    dom.window.document.body.innerHTML = FIGURE_CONTENT_XML;
    expect(parseFigure(dom.window.document.body.firstChild as Element, bodyState.schema)).toMatchSnapshot();
  });

  it('checks figure serialization', () => {
    const dom = parseXML(`<article><body>${FIGURE_CONTENT_XML}</body></article>`);
    const bodyState = createBodyState(dom.querySelector('body')!);
    expect(serializeFigure(bodyState.doc.firstChild!)).toMatchSnapshot();

  });
});
