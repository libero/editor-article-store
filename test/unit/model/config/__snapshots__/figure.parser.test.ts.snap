// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figure parsing spec checks empty figure parsing 1`] = `
Array [
  Object {
    "type": "figureTitle",
  },
  Object {
    "type": "figureLegend",
  },
  Object {
    "type": "figureAttribution",
  },
]
`;

exports[`figure parsing spec checks figure parsing 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "text": "Pellentesque porta interdum sodales (",
        "type": "text",
      },
      Object {
        "attrs": Object {
          "refId": "bib6",
          "refText": "Heissler and Sellers, 2016a",
        },
        "type": "refCitation",
      },
      Object {
        "text": "). Vestibulum",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "subscript",
          },
        ],
        "text": "id",
        "type": "text",
      },
      Object {
        "text": " interdum",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "superscript",
          },
        ],
        "text": "felis",
        "type": "text",
      },
      Object {
        "text": ". ",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": "Vestibulum",
        "type": "text",
      },
      Object {
        "text": " at ",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "underline",
          },
        ],
        "text": "viverra",
        "type": "text",
      },
      Object {
        "text": " lorem. ",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "strikethrough",
          },
        ],
        "text": "Aliquam Gravida",
        "type": "text",
      },
      Object {
        "text": " nulla orci, quis tristique tortor ornare at ",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "attrs": Object {
              "href": "https://www.biobam.com/",
            },
            "type": "link",
          },
        ],
        "text": "https://www.biobam.com/",
        "type": "text",
      },
      Object {
        "text": ". Pellentesque accumsan sed ipsum eget hendrerit (",
        "type": "text",
      },
      Object {
        "attrs": Object {
          "refId": "bib3",
          "refText": "Foley et al., 2018",
        },
        "type": "refCitation",
      },
      Object {
        "text": "). Pellentesque cursus, nibh eu mattis condimentum, massa libero faucibus enim, et tempor orci neque eget",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "superscript",
          },
        ],
        "text": "mi",
        "type": "text",
      },
      Object {
        "text": ". Phasellus lacus",
        "type": "text",
      },
      Object {
        "marks": Array [
          Object {
            "type": "subscript",
          },
        ],
        "text": "mi",
        "type": "text",
      },
      Object {
        "text": ", ultrices quis placerat id, dignissim ac erat.",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "type": "figureTitle",
  },
  Object {
    "type": "figureLegend",
  },
  Object {
    "type": "figureAttribution",
  },
]
`;

exports[`figure parsing spec checks figure serialization 1`] = `
<fig>
  <label>
    Figure 1.
  </label>
  <graphic
     xlink:href="IMAGE_URL"
    mime-subtype=""
    mimetype="image"
  />
  <caption>
    <title>
      Bacterial cells on the surface of a plant leaf.
    </title>
    <p>
      Pellentesque porta interdum sodales (
      <a
        class="citation"
        data-cit-type="reference"
        data-ref-id="bib6"
        data-ref-text="Heissler and Sellers, 2016a"
        href="#"
      />
      ). Vestibulum
      <sub>
        id
      </sub>
       interdum
      <sup>
        felis
      </sup>
      . 
      <bold>
        Vestibulum
      </bold>
       at 
      <underline>
        viverra
      </underline>
       lorem. 
      <sc>
        Aliquam Gravida
      </sc>
       nulla orci, quis tristique tortor ornare at 
      <ext-link
        ext-link-type="uri"
        xlink:href="https://www.biobam.com/"
      >
        https://www.biobam.com/
      </ext-link>
      . Pellentesque accumsan sed ipsum eget hendrerit (
      <a
        class="citation"
        data-cit-type="reference"
        data-ref-id="bib3"
        data-ref-text="Foley et al., 2018"
        href="#"
      />
      ). Pellentesque cursus, nibh eu mattis condimentum, massa libero faucibus enim, et tempor orci neque eget
      <sup>
        mi
      </sup>
      . Phasellus lacus
      <sub>
        mi
      </sub>
      , ultrices quis placerat id, dignissim ac erat.
    </p>
    <p />
  </caption>
</fig>
`;
